sdk/src/MSLoginSafariViewController.m

```
@@ -56,103 +48,128 @@ - (void)loginWithProvider:(NSString *)provider
                  animated:(BOOL)animated
                completion:(nullable MSClientLoginBlock)completion
 {
-    // SafariServices API is only available on iOS 9 or later, not iOS 8 or prior.
-    // When SafariServices is not avaiable, fallback to WebView based login in |MSLogin|
-    // for backward compatibility.
-    
-    if ([SFSafariViewController class]) {
-        [self safariViewControllerLoginWithProvider:provider
-                                        urlScheme:urlScheme
-                                         parameters:parameters
-                                         controller:controller
-                                           animated:animated
-                                         completion:completion];
-    }
-    else {
-        [self.webViewLogin loginWithProvider:provider
-                           parameters:parameters
-                           controller:controller
-                             animated:animated
-                           completion:completion];
-    }
+    dispatch_async(dispatch_get_main_queue(),^{
+        [self loginInternalWithProvider:provider
+                          urlScheme:urlScheme
+                         parameters:parameters
+                         controller:controller
+                           animated:animated
+                         completion:completion];
+    });
 }
 
 - (BOOL)resumeWithURL:(NSURL *)URL
 {
```

Current implementation already ensures safariViewController:dismissionViewControllerAnimated:completion runs on UI thread by dispatch_get_main_queue. I still agree with your suggestion despite of that. To make it fail early when not running from UI thread, checking it up-front whether it's UI thread would be a better approach.

---
In reply to: [96508405](https://github.com/Azure/azure-mobile-apps-ios-client/pull/122#discussion_r96508405) [](ancestors = 96508405)